### PR TITLE
checking collision between two sets of objects in the world

### PR DESCRIPTION
**Pull request by pradeepr-roboticist**
### Description

This PR is related to #2097 .
Here are the changes made:

A new virtual public function (checkCollisionBetweenObjectGroups) was added to collision_env.h.
This function provides the Boolean result of a collision check between objects of one group and another group. Each group is a vector of strings denoting the name of the objects in the group.

A concrete implementation is provided in "collision_env_fcl.cpp" for the new function.
The implementation instantiates a FCL CollisionManager for each group and populates the collision manager with the corresponding objects.
Finally, a collision query is made to the CollisionManager and the result is returned.

A test case is also provided in "test_fcl_env.cpp".
 
### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [x] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [x] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: MIGRATION.md [hunk 1]
```diff
@@ -15,6 +15,7 @@ API changes in MoveIt releases
 - Requests to `get_planning_scene` service without explicitly setting "components" now return full scene
 - `moveit_ros_planning` no longer depends on `moveit_ros_perception`
 - `CollisionRobot` and `CollisionWorld` are combined into a single `CollisionEnv` class. This applies for all derived collision checkers as `FCL`, `ALL_VALID`, `HYBRID` and `DISTANCE_FIELD`. Consequently, `getCollisionRobot[Unpadded] / getCollisionWorld` functions are replaced through a `getCollisionEnv` in the planning scene and return the new combined environment. This unified collision environment provides the union of all member functions of `CollisionRobot` and `CollisionWorld`. Note that calling `checkRobotCollision` of the `CollisionEnv` does not take a `CollisionRobot` as an argument anymore as it is implicitly contained in the `CollisionEnv`.
+- `CollisionEnv` provides a function `checkCollisionBetweenObjectGroups` that allows collision detection between two sets of objects that are already part of the world ([2097](https://github.com/ros-planning/moveit/issues/2097)).
```
Comment (v4hn) [inline MIGRATION.md:18]: No need for this entry. `MIGRATION.md` is meant to include only API-incompatible changes.

File: moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp [hunk 2]
```diff
@@ -309,6 +309,45 @@ void CollisionEnvFCL::checkRobotCollisionHelper(const CollisionRequest& req, Col
...
+  auto manager1 = std::make_unique<fcl::DynamicAABBTreeCollisionManagerd>();
+  auto manager2 = std::make_unique<fcl::DynamicAABBTreeCollisionManagerd>();
+
+  // Populate managers with the FCL objects corresponding entries in object_groups
+  for (auto it = object_group1.begin(); it != object_group1.end(); ++it)
+  {
+    auto const& object_name = *it;
+    if (fcl_objs_.count(object_name) == 0)
+      ROS_ERROR_STREAM_NAMED("collision_detection.fcl",
+                             "FCL object by the name: " << object_name << " does not exist.");
```
Comment (v4hn) [inline moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp:324]:
```suggestion
                             "FCL object by the name: '" << object_name << "' does not exist.");
```

File: moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp [hunk 3]
```diff
@@ -309,6 +309,45 @@ void CollisionEnvFCL::checkRobotCollisionHelper(const CollisionRequest& req, Col
   }
 }
 
+bool CollisionEnvFCL::checkCollisionBetweenObjectGroups(std::vector<std::string> const& object_group1,
+                                                        std::vector<std::string> const& object_group2) const
```
Comment (v4hn) [inline moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp:313]: I would expect the possibility to specify only *one* set of objects and check against everything else.

**Comment by v4hn** [comment]
Also,

- I think it makes sense to provide the same interface for the `Distance` requests too. If we can test for collisions, we might as well allow to compute distances.
- In many places, the low-level APIs try to avoid the overhead of lots of string lookups. Is there a reasonable way to collect pointers to the fcl objects beforehand and reuse them across multiple requests?

File: moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp (hunk 3 above)
Comment (pradeepr-roboticist) [reply moveit_core/collision_detection_fcl/src/collision_env_fcl.cpp:313]: That sounds like a good idea. I will look into that as well.

**Comment by pradeepr-roboticist** [comment]
> Also,
> 
> * I think it makes sense to provide the same interface for the `Distance` requests too. If we can test for collisions, we might as well allow to compute distances.

I will look into this feature as well.

> * In many places, the low-level APIs try to avoid the overhead of lots of string lookups. Is there a reasonable way to collect pointers to the fcl objects beforehand and reuse them across multiple requests?

Sure. I understand the performance issue with string lookups. I'll find a way to replace it with FCL object pointers.

**Comment by felixvd** [comment]
@pradeepr-roboticist Any progress on this?

**Comment by felixvd** [comment]
I'll close this for lack of activity. Feel free to reopen after addressing the changes.

**Pull request closed without merging**